File: backend2/workers/worker.py
import redis
import sys
import json
import os
import subprocess
import time
from dotenv import load_dotenv

load_dotenv()

r = redis.Redis(
    host=os.getenv("REDIS_HOST"),
    port=int(os.getenv("REDIS_PORT")),
    db=int(os.getenv("REDIS_DB")),
    decode_responses=True
)

QUEUE = os.getenv("QUEUE_NAME", "network-queue")

SCRIPT = "/home/laxmikant/Virtual-Cyber-Labs/backend2/scripts/create_user_bridges.py"
PYTHON = "/home/laxmikant/Virtual-Cyber-Labs/backend2/venv/bin/python3"


def process(user_id):
    print(f"[+] Creating network for user {user_id}", flush=True)

    subprocess.run(
        ["sudo", PYTHON, SCRIPT, str(user_id)],
        check=True
    )

    print(f"[✓] Done user {user_id}", flush=True)


def start_worker():
    print("[WORKER] Started...", flush=True)

    while True:
        try:
            _, job = r.brpop(QUEUE)
            data = json.loads(job)

            user_id = data.get("user_id")

            if not user_id:
                print("[SKIP] Invalid job:", data, flush=True)
                continue

            print(f"[JOB] Processing user {user_id}", flush=True)

            process(user_id)

        except Exception as e:
            print("[ERROR]", e, flush=True)
            time.sleep(2)


if __name__ == "__main__":
    start_worker()

File: backend/workers/worker.js
const { Worker } = require("bullmq");
const IORedis = require("ioredis");
const { spawn } = require("child_process");

const connection = new IORedis({
  host: "127.0.0.1",
  port: 6379,
  maxRetriesPerRequest: null,
});

const worker = new Worker(
  "network-queue",
  async (job) => {
    console.log("📦 JOB RECEIVED:", job.data);

    const { userId } = job.data;

    return new Promise((resolve, reject) => {
      const process = spawn("sudo", [
        "python3",
        "/home/laxmikant/Virtual-Cyber-Labs/backend/scripts/create_user_bridges.py",
        userId.toString(),
      ]);

      process.stdout.on("data", (data) => {
        console.log(`[PYTHON]: ${data}`);
      });

      process.stderr.on("data", (data) => {
        console.error(`[PYTHON ERROR]: ${data}`);
      });

      process.on("close", (code) => {
        if (code === 0) {
          console.log(`✅ Network created for user ${userId}`);
          resolve();
        } else {
          reject(new Error(`Exit code ${code}`));
        }
      });
    });
  },
  { connection }
);

worker.on("completed", (job) => {
  console.log(`✅ Job completed: ${job.id}`);
});

worker.on("failed", (job, err) => {
  console.log(`❌ Job failed: ${job.id}`, err.message);
});

console.log("🔥 Worker started...");